iroh-sync/src/store/fs.rs

```diff
@@ -273,75 +230,68 @@ impl super::Store for Store {
 }
 
 impl Store {
-    fn get_latest_by_key(
+    fn get_by_key(
         &self,
         namespace: NamespaceId,
         key: impl AsRef<[u8]>,
-    ) -> Result<RangeLatestIterator<'_>> {
-        RangeLatestIterator::namespace(
+    ) -> Result<RangeIterator<'_>> {
+        RangeIterator::namespace(
             &self.db,
             &namespace,
             RangeFilter::Key(key.as_ref().to_vec()),
         )
     }
-
-    fn get_latest_by_prefix(
-        &self,
-        namespace: NamespaceId,
-        prefix: impl AsRef<[u8]>,
-    ) -> Result<RangeLatestIterator<'_>> {
-        RangeLatestIterator::namespace(
-            &self.db,
-            &namespace,
-            RangeFilter::Prefix(prefix.as_ref().to_vec()),
-        )
-    }
-
-    fn get_latest(&self, namespace: NamespaceId) -> Result<RangeLatestIterator<'_>> {
-        RangeLatestIterator::namespace(&self.db, &namespace, RangeFilter::None)
-    }
-
-    fn get_all_by_key_and_author<'a, 'b: 'a>(
-        &'a self,
-        namespace: NamespaceId,
-        author: AuthorId,
-        key: impl AsRef<[u8]> + 'b,
-    ) -> Result<RangeAllIterator<'a>> {
-        let start = (namespace.as_bytes(), author.as_bytes(), key.as_ref());
-        let end = (namespace.as_bytes(), author.as_bytes(), key.as_ref());
-        RangeAllIterator::with_range(
-            &self.db,
-            |table| table.range(start..=end),
-            RangeFilter::None,
-        )
+    fn get_by_author(&self, namespace: NamespaceId, author: AuthorId) -> Result<RangeIterator<'_>> {
+        let author = author.as_bytes();
+        let start = (namespace.as_bytes(), author, &[][..]);
+        let mut author2 = *author;
+        // move to next possible author, which marks the end of the range
+        // TODO: is there a better way to define the range end?
+        for char in author2.iter_mut().rev() {
+            if *char != 255 {
+                *char += 1;
+                break;
+            }
+        }
+        let end = (namespace.as_bytes(), &author2, &[][..]);
+        RangeIterator::with_range(&self.db, |table| table.range(start..end), RangeFilter::None)
     }
 
-    fn get_all_by_key(
+    fn get_by_author_and_prefix(
         &self,
         namespace: NamespaceId,
-        key: impl AsRef<[u8]>,
-    ) -> Result<RangeAllIterator<'_>> {
-        RangeAllIterator::namespace(
-            &self.db,
-            &namespace,
-            RangeFilter::Key(key.as_ref().to_vec()),
-        )
+        author: AuthorId,
+        prefix: impl AsRef<[u8]>,
+    ) -> Result<RangeIterator<'_>> {
+        let author = author.as_bytes();
+        let start = (namespace.as_bytes(), author, prefix.as_ref());
+        let mut end_prefix = prefix.as_ref().to_vec();
+        // move to next possible key after the prefix, which marks the end of the range
+        // TODO: is there a better way to define the range end?
+        for char in end_prefix.iter_mut().rev() {
```

Comment:
this won't work for a key that is all 255

Reply:
Yup. To solve this, we need to add a seperator byte between author and key. Maybe also between namespace and author (or not if we're positive that `[255; 32]` never occurs in practice as èd25591` pubkey.

Reply:
Right. For an all-255 prefix, we should increment the last byte of the author then instead.

Reply:
Fixed and added a unit test